### PR TITLE
Add basic type hints to the whole package

### LINKED ISSUE #344: Pylance reports no stub files
I have Pylance configured with `"python.analysis.typeCheckingMode": "strict"` and it's complaining that it doesn't find typing stubs:

image

Would you accept a PR adding type-hints to the whole package?

### PR DESCRIPTION
**Pull request by sebastian-correa**
# Summary
This PR adds basic type-hints to the whole package. Closes #344.

# Questions
1. Why are there 2 imports _at the bottom_ of `utils.py`? I tried removing the unused `settings` import and moving the `FakeUserAgentError` to the top and tests passed.
2. There are checks to see if `self.fallback is None` [[1](https://github.com/fake-useragent/fake-useragent/blob/88645b5fdbd1624229cdee6f7b809d56627d5d61/src/fake_useragent/fake.py#L127), [2](https://github.com/fake-useragent/fake-useragent/blob/88645b5fdbd1624229cdee6f7b809d56627d5d61/src/fake_useragent/fake.py#L180)] but, by construction, it can only be a `str` [[3](https://github.com/fake-useragent/fake-useragent/blob/88645b5fdbd1624229cdee6f7b809d56627d5d61/src/fake_useragent/fake.py#L60)]. Shuold we type-hint it `str | None` and change the `assert`ion or should the code checking against `None` be changed?
3. What is the "system" key in the data, so that we can add better docstrings?
4. What's the point of `str_types`? I currently type-hint `fallback` as `set[str]` when it should really be a `set` of members of that `tuple`. I can look into it if we want to keep that :).
5. The `str, Enum`s I added might be a bit much. I like them because they are self-documenting and because they inherit from `str, Enum`, members are `str`s, so that's transparent to the user. However, it might be too much of a burden to remember to add elements to the `Enum` when changing the `JSON`. What do you think?

# Notes
If this PR is merged, I wouldn't do a release yet. There are some changes I'd like to propose regarding the defaults and asserts, but I'll be opening new issues for those so you can decide what you prefer!

### CODE REVIEW

**Comment by melroy89** [comment]
_Updated the branch, so it's in sync again with the latest main_

File: src/fake_useragent/utils.py
Comment (melroy89) [inline src/fake_useragent/utils.py:30]:
```suggestion
    os: OS
```

I think you then also want to use the class?

**Comment by melroy89** [comment]
Regarding your first 2 questions...

1. Yea, this is some legacy code. I cleaned it up in PR: https://github.com/fake-useragent/fake-useragent/pull/351
2. Also legacy issue, also cleaned it up in the same PR https://github.com/fake-useragent/fake-useragent/pull/351

Please rebase :) and fix conflicts.

**Comment by melroy89** [comment]
_Fixed the file conflicts for you_

**Comment by melroy89** [comment]
Now, let me try to answer the remaining 3 answers you had.

> What is the "system" key in the data, so that we can add better docstrings?

This is the operating kind of string.. Also legacy code. It was part of: https://github.com/fake-useragent/fake-useragent/pull/290 .. Maybe we should not use it anymore.. Or clean up the system string better. But for now it seems to be a combination of the browser name + OS. For example: `Edge 121.0 Windows`

Maybe it doesn't add any value anymore.. 

> What's the point of str_types? I currently type-hint fallback as set[str] when it should really be a set of members of that tuple. I can look into it if we want to keep that :).

To be fully honest, this is code not written by me (but before I took over maintainer role). Anyhow, `str_types`  is checked on line 67 in fake.py to validate the input `safe_attrs` for a valid type, which is  a tuple of str == `(str,)`... If that is all correct it will set `self.safe_attrs`. Then in the `__getattr__` method there is piece of code that check the incoming attr string against the `self.safe_attrs` and if the input is listed in that tuple, then it will call the super method of `__getattribute__`:

```py
        if attr in self.safe_attrs:
            return super(UserAgent, self).__getattribute__(attr)
```

It's explained here in the note: https://github.com/fake-useragent/fake-useragent?tab=readme-ov-file#notes. With an example:

```py
ua = UserAgent(safe_attrs=('__injections__',))
```

> The str, Enums I added might be a bit much. I like them because they are self-documenting and because they inherit from str, Enum, members are strs, so that's transparent to the user. However, it might be too much of a burden to remember to add elements to the Enum when changing the JSON. What do you think?

I do agree that it will create more work to keep the data in-sync with the JSON. Which will most likely soon update (again) again (again) as well. However, I do like the transparency it gives to the clients. 

That being said, to be honest.. We can't grantee all the output.. It could also become `Windows NT 5.1` or `Windows NT 6.1` soon.. These strings are constantly changing

**Comment by sebastian-correa** [comment]
I'll look at your comments tonight (I'm GMT-3), but thanks for the code updates and explanations you provided!

**Comment by melroy89** [comment]
> I'll look at your comments tonight (I'm GMT-3), but thanks for the code updates and explanations you provided!

Thanks in advance! FYI I'm in GMT+2 (while it's still daylight saving time).

**Comment by sebastian-correa** [comment]
> > I'll look at your comments tonight (I'm GMT-3), but thanks for the code updates and explanations you provided!
> 
> Thanks in advance! FYI I'm in GMT+2 (while it's still daylight saving time).

Never visited Tiel, but I LOVED my time in Eindhoven (1 semester of exchange at TU/e). I really miss it there, and Jumbo's peach greek yogurt (what a delight that cheap yogurt is, my god 🤣).

**Comment by melroy89** [comment]
Here you have the image as a souvenir:

image

**Comment by sebastian-correa** [comment]
> Maybe it doesn't add any value anymore..

Maybe not, but changing it would require a major version bump and that might complicate things. We can think about how to deprecate it "gradually" with deprecation warnings maybe in a separate issue. For now, I suggest we keep the basic doc I added.

> Anyhow, str_types is checked on line 67 in fake.py to validate the input safe_attrs for a valid type, (...)

Agreed on where it's used and how. Question is _why_ does it exist? By the looks of #151, it seems like this was there to support Python 2 versions. I propose we get rid of the constant and just check against `str`. What do you think? Otherwise, I can just change `self.safe_attrs` to be a `set[Union[str_types]]`, which I think should work fine but seems pointless.

> That being said, to be honest.. We can't grantee all the output.. It could also become Windows NT 5.1 or Windows NT 6.1 soon.. These strings are constantly changing

Sounds reasonable. Let's roll those back and keep just `str`s for now. I mostly wanted the `Browser` one to make typing the `__getattr__` easier, but I think there's a way around that which I can try in a future PR if this one gets merged.

**Comment by melroy89** [comment]
> Agreed on where it's used and how. Question is _why_ does it exist? By the looks of #151, it seems like this was there to support Python 2 versions. I propose we get rid of the constant and just check against `str`. What do you think? Otherwise, I can just change `self.safe_attrs` to be a `set[Union[str_types]]`, which I think should work fine but seems pointless.

No this safe attribute code was also prio to there.. before #151... I just refactor the code a bit, but again the `safe_attrs=tuple(),` input on the class was already there from the previous maintainer of this project.

And again the [README.md](https://github.com/fake-useragent/fake-useragent/blob/main/README.md) says:

> If you need to safe some attributes from overriding them in UserAgent by __getattr__ method use safe_attrs you can pass there attributes names. At least this will prevent you from raising FakeUserAgentError when attribute not found.
> 
> For example, when using fakeuseragent with injections <https://github.com/tailhook/injections> you need to:
> 
> from fake_useragent import UserAgent
> 
> ua = UserAgent(safe_attrs=('__injections__',))
> 
> Please, do not use if you don't understand why you need this. This is magic for rarely extreme case.

**Review by melroy89 — APPROVED** [review]

**Comment by melroy89** [comment]
LGTM now.. Anything to add? I will wait for your approval as well. And yes there will be a new release soonish. But I feel free to create some additional PRs if needed. I also want to get this new script in: https://github.com/fake-useragent/fake-useragent/pull/356.. And update the browsers.json to avoid duplicate UA strings and more unique strings in general.

**Comment by sebastian-correa** [comment]
I would merge as-is as the code will be slightly refined in my upcoming planned PRs.

**Pull request merged**
